### PR TITLE
Space Objects : Share options

### PR DESCRIPTION
**Pull request by kumarpalsinh25**
This PR contains basic work around of Share Space Object options. At the moment sharing space object with new boost is support.

This PR also contain basic work around of Deep-linking work. At the moment, Space Object QR option is supported.

Video Reference:

https://github.com/user-attachments/assets/d5fe1f30-8d40-420f-a8e3-11246952bc81

--- 
UI Reference when all option visible:

https://github.com/user-attachments/assets/7528ae8e-12b7-472a-acbd-32a51d7a1a8a

### CODE REVIEW

File: app/lib/common/widgets/share/action/share_space_object_action.dart [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+  ObjectType objectType,
+  String objectId,
+});
+
+typedef FileDetails = ({
+  File file,
+  String? mimeType,
+});
+
+Future<void> openShareSpaceObjectDialog({
```
Comment (kumarpalsinh25) [inline app/lib/common/widgets/share/action/share_space_object_action.dart:27]: Generic action for spaceObject share option.

File: app/lib/common/widgets/share/widgets/attach_options.dart [hunk 2]
```diff
@@ -0,0 +1,124 @@
+import 'package:acter/common/themes/colors/color_scheme.dart';
+import 'package:atlas_icons/atlas_icons.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter_gen/gen_l10n/l10n.dart';
+
+class AttachOptions extends StatelessWidget {
```
Comment (kumarpalsinh25) [inline app/lib/common/widgets/share/widgets/attach_options.dart:6]: Generic `AttachOptions` UI Widget which is independent of data part.

File: app/lib/common/widgets/share/widgets/external_share_options.dart [hunk 3]
```diff
@@ -0,0 +1,135 @@
+import 'package:flutter/material.dart';
+import 'package:phosphor_flutter/phosphor_flutter.dart';
+import 'package:flutter_gen/gen_l10n/l10n.dart';
+
+class ExternalShareOptions extends StatelessWidget {
```
Comment (kumarpalsinh25) [inline app/lib/common/widgets/share/widgets/external_share_options.dart:5]: Generic `ExternalShareOptions` UI Widget which is independent of data part.

File: app/lib/common/widgets/share/widgets/file_share_options.dart [hunk 4]
```diff
@@ -0,0 +1,92 @@
+import 'package:flutter/material.dart';
+import 'package:phosphor_flutter/phosphor_flutter.dart';
+import 'package:flutter_gen/gen_l10n/l10n.dart';
+
+class FileShareOptions extends StatelessWidget {
```
Comment (kumarpalsinh25) [inline app/lib/common/widgets/share/widgets/file_share_options.dart:5]: Generic `FileShareOptions` UI Widget which is independent of data part.

File: app/lib/features/news/news_utils/news_utils.dart [hunk 5]
```diff
@@ -51,18 +52,25 @@ class NewsUtils {
...
+  static void addTextSlide({
+    required WidgetRef ref,
+    NewsReferencesModel? newsReferencesModel,
+  }) {
     final clr = getRandomElement(newsPostColors);
     NewsSlideItem textSlide = NewsSlideItem(
       type: NewsSlideType.text,
       text: '',
       backgroundColor: clr,
+      newsReferencesModel: newsReferencesModel,
```
Comment (kumarpalsinh25) [inline app/lib/features/news/news_utils/news_utils.dart:64]: Add `newsReferencesModel` widget if available.

File: app/lib/features/news/pages/add_news_page.dart [hunk 6]
```diff
@@ -24,10 +25,12 @@ const addNewsKey = Key('add-news');
 
 class AddNewsPage extends ConsumerStatefulWidget {
   final String? initialSelectedSpace;
+  final NewsReferencesModel? newsReferencesModel;
```
Comment (kumarpalsinh25) [inline app/lib/features/news/pages/add_news_page.dart:28]: Accept `newsReferencesModel` widget as option argument while creating boost.

File: .changes/2390-space-object-share.md [hunk 7]
```diff
@@ -0,0 +1,2 @@
+- [New] : Now you can directly share space objects (Pin, Event and TaskList) to new Boost from their detail page.
+- [New] : Now space objects can accessible by their own unique QR code.
```
Comment (gnunicorn) [inline .changes/2390-space-object-share.md:2]:
```suggestion
- [New] : You can now share any object via its own unique QR code.
```

File: app/lib/common/widgets/share/action/share_space_object_action.dart [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+        ),
+      ),
+    );
+  }
+
+  Widget attachmentOptionsUI(BuildContext context) {
+    String spaceId = spaceObjectDetails!.spaceId;
+    ObjectType objectType = spaceObjectDetails!.objectType;
+    String objectId = spaceObjectDetails!.objectId;
+
```
Comment (gnunicorn) [inline app/lib/common/widgets/share/action/share_space_object_action.dart:81]: hmm... I _know_ from the code that this is fine, but in the past `!` has bitten us in that when code around it is updated the assumption breaks and all we have is a rather undescriptive error in sentry. Can we refactor this to not have `!` here? e.g. by passing the object as a parameter?

File: app/lib/common/widgets/share/action/share_space_object_action.dart [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+          extra: newsRefType != null
+              ? NewsReferencesModel(type: newsRefType, id: objectId)
+              : null,
+        );
+      },
+    );
+  }
+
+  Widget externalShareOptionsUI(BuildContext context) {
+    String spaceId = spaceObjectDetails!.spaceId;
```
Comment (gnunicorn) [inline app/lib/common/widgets/share/action/share_space_object_action.dart:98]: same as above

File: app/lib/common/widgets/share/widgets/external_share_options.dart [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+                ),
+              if (onTapSignal != null)
+                shareToItemUI(
+                  name: lang.signal,
+                  iconData: PhosphorIcons.chat(),
+                  color: Colors.deepPurpleAccent,
+                  onTap: onTapSignal,
+                ),
+              if (onTapWhatsApp != null)
+                shareToItemUI(
```
Comment (gnunicorn) [inline app/lib/common/widgets/share/widgets/external_share_options.dart:74]: we should also check that whatsApp is available on the system in general, no? same for all the others, too ...

File: app/lib/common/widgets/share/widgets/external_share_options.dart [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+                  name: lang.copyLink,
+                  iconData: PhosphorIcons.link(),
+                  color: Colors.blueGrey,
+                  onTap: onTapCopy,
+                ),
+              if (onTapSignal != null)
+                shareToItemUI(
+                  name: lang.signal,
+                  iconData: PhosphorIcons.chat(),
```
Comment (gnunicorn) [inline app/lib/common/widgets/share/widgets/external_share_options.dart:69]: let's put a fixme in here that we want to include signals actual logo in the source

**Review by gnunicorn — APPROVED** [review]

File: app/lib/common/widgets/share/action/share_space_object_action.dart (hunk 8 above)
Comment (kumarpalsinh25) [reply app/lib/common/widgets/share/action/share_space_object_action.dart:81]: Yes sure.

File: app/lib/common/widgets/share/widgets/external_share_options.dart (hunk 10 above)
Comment (kumarpalsinh25) [reply app/lib/common/widgets/share/widgets/external_share_options.dart:74]: Yes, this is in my TO-DO List. Planing to add it next upcoming PR.

File: app/lib/common/widgets/share/widgets/external_share_options.dart (hunk 11 above)
Comment (kumarpalsinh25) [reply app/lib/common/widgets/share/widgets/external_share_options.dart:69]: Added Icon.

Screenshot 2024-12-04 at 12 46 38 PM

**Pull request merged**